plugins/SlicerT/SlicerTWaveform.cpp

```diff
@@ -0,0 +1,335 @@
+/*
+ * SlicerTWaveform.cpp - slice editor for SlicerT
+ *
+ * Copyright (c) 2023 Daniel Kauss Serna <daniel.kauss.serna@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SlicerTWaveform.h"
+
+#include "SlicerT.h"
+#include "SlicerTView.h"
+#include "embed.h"
+
+namespace lmms {
+
+namespace gui {
+
+SlicerTWaveform::SlicerTWaveform(int w, int h, SlicerT* instrument, QWidget* parent)
+	: QWidget(parent)
+	,
+	// calculate sizes
+	m_width(w)
+	, m_height(h)
+	, m_seekerWidth(w - m_seekerHorMargin * 2)
+	, m_editorHeight(h - m_seekerHeight - m_middleMargin)
+	, m_editorWidth(w)
+	,
+
+	// create pixmaps
+	m_sliceArrow(PLUGIN_NAME::getIconPixmap("slide_indicator_arrow"))
+	, m_seeker(QPixmap(m_seekerWidth, m_seekerHeight))
+	, m_seekerSlicerTWaveform(QPixmap(m_seekerWidth, m_seekerHeight))
+	, m_sliceEditor(QPixmap(w, m_editorHeight))
+
+	// references to instrument vars
+	, m_slicerTParent(instrument)
+{
+	// window config
+	setFixedSize(m_width, m_height);
+	setMouseTracking(true);
+
+	// draw backgrounds
+	m_sliceEditor.fill(s_SlicerTWaveformBgColor);
+	m_seekerSlicerTWaveform.fill(s_SlicerTWaveformBgColor);
+
+	// connect to playback
+	connect(instrument, SIGNAL(isPlaying(float, float, float)), this, SLOT(isPlaying(float, float, float)));
+	connect(instrument, SIGNAL(dataChanged()), this, SLOT(updateUI()));
```

Comment:
The `SIGNAL` and `SLOT` macros should be avoided in modern Qt

Reply:
Done